gdal/doc/source/development/rfc/rfc76_ogrpythondrivers.rst

```diff
@@ -0,0 +1,397 @@
+            if self.attribute_filter is not None or \
+            self.geometry_filter is not None:
+                return super(Layer, self).feature_count(force_computation)
+
+            return self.count
+
+        # Optional. Should generally be exclusive with feature_count()
+        def unfiltered_feature_count(self, force_computation):
+            assert self.attribute_filter is None and self.geometry_filter is None
+            return self.count
+
+        # Required
+        def reset_reading(self):
+            self.iter = 0
+
+        # Optional, unless next_raw_feature is defined.
+        # Must then honour self.attribute_filter and self.geometry_filter
+        # attribute_filter will be None or a OGR SQL WHERE clause
+        # geometry_filter will be None or [minx, miny, maxx, maxy]
+        # def next_feature(self):
+        #     return {}
+
+        # Required, unless nextFeature is defined.
+        def next_raw_feature(self):
+            if self.iter == self.count:
+                return None
+            self.iter += 1
+
+            # A GeoJSON Feature object might also be returned if there is a
+            # single geometry field
+            return {'my_fid': self.iter,
+                    'boolField': True,
+                    'intField': self.iter + 1,
+                    'int64Field': 1234567890123,
+                    'realField': 1.23,
+                    'strField': 'foo',
+                    'strNullField': None,
+                    'binaryField': b'\x01\x00\x02',
+                    'datetimeField': '2017-04-26T12:34:56.789Z',
+                    'geomField': 'POINT(2 49)'}
+
+
+    class Dataset(GDALDataset):
+
+        def __init__(self, filename):
+            self.layers = [Layer()]
+            pass
+
+        # Optional, called on native object destruction
+        def __del__(self):
+            pass
+
+        # Optional
+        def metadata(self, domain):
+            if domain is None:
+                return {'foo': 'bar'}
+            return None
+
+        # Required, unless a layers attribute is set in __init__
+        # def layer_count(self):
+        #     return len(self.layers)
+
+        # Required, unless a layers attribute is set in __init__
+        # def layer(self, idx):
+        #     return self.layers[idx]
+
+
+    class Driver(GDALDriver):
+
+        # Required
+        def identify(self, filename, first_bytes, open_flags):
+            return filename == 'MYPLUGIN:'
+
+        # Required
+        def open(self, filename, first_bytes, open_flags):
+            if not self.identify(filename, first_bytes, open_flags):
+                return None
+            return Dataset(filename)
+
+
+    # Required
+    def create_driver():
+        return Driver()
+
+
+Limitations and scope
+---------------------
+
+- Vector and read-only for now. This could later be extended of course.
+
+- No connection between the Python code of the plugin and the OGR Python API 
```

Comment:
Simply, the new one is "GDAL in Python"  and the existing one is "Python in GDAL" :-)

Reply:
@mloskot or in other words, extending Python vs embedding Python: https://docs.python.org/3/extending/. I don't think it's a great idea for GDAL to promote both, but maybe that's just where GDAL is in the software universe, and the Python docs do cover the topic of extending embedded Python: https://docs.python.org/3/extending/embedding.html#extending-embedded-python.

Reply:
Good point. Having written all three (extend, embed, extend embedded incl. extending with plain Python modules) and I don't think this proposal falls that into canonical embedded (and extend embedded) with Python.

Reply:
> extending Python vs embedding Python:

Discovering the taxonomy of those Python techniques :-) As far as I understand the docs, the techniques I use here are embedding Python (to load the user .py script, and the empty definitions of the OGRLayer, GDALDataset and GDALDriver classes) and extending python to define a _gdal_base module (internal detail) that has a layer_featureCount() method that is called by OGRLayer.feature_count() default implementation

> One can imagine scenarios where someone tries to use a Python driver in GDAL (embedded Python) and the Python driver uses the GDAL bindings (extending Python) and a VRT with Python functions (embedded Python again), and so on. The possibilities for bugs are endless.

I just tested that in https://gist.github.com/rouault/1dcc5a13583ba2ab419cf4a8f05b1ccc. It works fine from ogrinfo command line. Works fine also from a Python console doing "from osgeo import ogr; ds = ogr.Open('MYPLUGIN:')"  (so Python (console) calling C++ calling Python (OGR Python driver) calling C++ calling Python (embedded Python in VRT))  . I agree there might be some fragility though. Hence the limited scope to experimentation for now to see how it flyes with users.

> Or, it is to allow import x where x is not necessarily installed and available for embedded Python...

Yes, the Python drivers need to import gdal_base that has the base objects and some hidden magic to interact with the native code. This makes deployment easier rather than having a real .py + .so small module extending Python (like the class SWIG bindings do)

Reply:
@rouault 
> As far as I understand the docs, the techniques I use here are embedding Python (to load the user .py script, and the empty definitions of the OGRLayer, GDALDataset and GDALDriver classes) and extending python to define a _gdal_base module (internal detail)  ...

In my opinion, the user-defined subclass of `Driver`, `OGRLayer` pre-defined classes are also an extension to the embedded Python.

The way you are proposing it is slightly more complex than canonical way of embedding Python:
it is not linking  Python library at build time in order to embed the Python interpreter and API, but actually embedding Python at run-time. I'd call it an extremely delayed or lazy embedding :)

I would not be surprised if some readers of the RFC and users (especially not experienced w/ embedding) of the machinery are a bit confused asking on the list how it works, because it is not a typical way of embedding.

Reply:
> The possibilities for bugs are endless.

PDAL both extends and embeds Python. This statement is a fact. It is also really useful to do both things sometimes.

Reply:
few words added about that